### PR TITLE
Add translations for zh-CN locale

### PR DESCRIPTION
**Pull request by yfdyh000**
A part of #116.

### CODE REVIEW

File: macros/AppsLandingRightSide.ejs [hunk 1]
```diff
@@ -74,6 +74,40 @@ switch(env.locale) {
 <p><span class="alllinks"><a rel="external" href="http://www.catb.org/~esr/faqs/smart-questions.html"><em>অনলাইন আচরণবিধি</em> ভুলে গেলে চলবে না...</a></span></p><%
 
     break;
+    case "zh-CN":
+
+%><h5 class="Tools" id="Tools" name="Tools">面向应用开发者的工具</h5>
```
Comment (jwhitlock) [inline macros/AppsLandingRightSide.ejs]: 👍 , tested on copy of https://developer.mozilla.org/en-US/Apps/Fundamentals/Games.  This macro is a good candidate for deletion.

File: macros/ArticlesByTag.ejs [hunk 2]
```diff
@@ -16,6 +16,7 @@
 var l10n = {
     noArticles : mdn.localString({
         "en-US": "There have been no articles written yet. <a href=\"/en-US/docs/MDN/Getting_started\">Please consider contributing</a>.",
+        "zh-CN": "这些文章尚未撰写。<a href=\"/en-US/docs/MDN/Getting_started\">欢迎贡献</a>。",
```
Comment (jwhitlock) [inline macros/ArticlesByTag.ejs]: Translation looks good, but the link should go to [/zh-CN/docs/MDN/Getting_started](https://developer.mozilla.org/zh-CN/docs/MDN/Getting_started). This macro is a good candidate for deletion, it is only used on [9, non-English pages](https://developer.mozilla.org/en-US/search?locale=%2A&kumascript_macros=ArticlesByTag&topic=none).

File: macros/AvailableInWorkers.ejs [hunk 3]
```diff
@@ -3,6 +3,7 @@ var locale = env.locale;
 
 var text = mdn.localString({
   "en-US": "This feature is available in <a href=\"/" + locale + "/docs/Web/API/Web_Workers_API\">Web Workers</a>.",
+  "zh-CN": "此特性在 <a href=\"/" + locale + "/docs/Web/API/Web_Workers_API\">Web Worker</a> 中可用。",
```
Comment (jwhitlock) [inline macros/AvailableInWorkers.ejs]: This works (tested on copy of [WebSockets/Writing_WebSocket_client_applications](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications).  Should ``Web Worker`` be ``Web Workers``, like the other translations?

File: macros/B2GMain.ejs [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 var prevStr = mdn.localString({
     "en-US": "« Back to Boot to Gecko",
+    "zh-CN": "« 返回 Boot to Gecko",
```
Comment (jwhitlock) [inline macros/B2GMain.ejs]: This doesn't break the page, tested on a copy of [B2G_OS/Automated_testing](https://developer.mozilla.org/en-US/docs/Archive/B2G_OS/Automated_testing).  However, the macro has been altered to create no output, so the translation is unnecessary.

File: macros/B2GOnlyHeader2.ejs [hunk 5]
```diff
@@ -37,6 +43,8 @@ var secString = {
     privileged: mdn.localString({
         "en-US" : 'This API is available on <a href="/en-US/docs/Mozilla/Firefox_OS">Firefox OS</a> for ' +
                   '<a href="/en-US/docs/Mozilla/Firefox_OS/Security/Application_security#App_Types">privileged or certified applications</a> only.',
+        "zh-CN" : '此 API 仅在 <a href="/zh-CN/docs/Mozilla/Firefox_OS">Firefox OS</a> 的' +
+                  '<a href="/zh-CN/docs/Mozilla/Firefox_OS/Security/Application_security#App_Types">高权限或已认证应用程序</a>上可用。',
```
Comment (jwhitlock) [inline macros/B2GOnlyHeader2.ejs]: 👍 tested on a translation of [Archive/B2G_OS/API/UDPSocket](https://developer.mozilla.org/en-US/docs/Archive/B2G_OS/API/UDPSocket)

File: macros/CSS_key_concepts.ejs [hunk 6]
```diff
@@ -75,6 +75,16 @@ switch(env.locale) {
...
+ <a href="/zh-CN/docs/Web/CSS/Comments" title="注释">注释</a>,
+ <a href="/zh-CN/docs/Web/CSS/Specificity" title="优先级">优先级</a>和<a href="/zh-CN/docs/Web/CSS/inheritance" title="继承">继承</a>,
+ the <a href="/zh-CN/docs/Web/CSS/CSS_Box_Model/Introduction_to_the_CSS_box_model" title="盒模型">盒模型</a>,
+ <a href="/zh-CN/docs/Web/CSS/Layout_mode" title="CSS 布局模式">布局模式</a>和<a href="/zh-CN/docs/Web/Guide/CSS/Visual_formatting_model" title="视觉格式化模型">视觉格式化模型</a>，以及<a href="/zh-CN/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing" title="外边距合并">外边距合并</a>，或者<a href="/zh-CN/docs/Web/CSS/initial_value" title="初始值">初始</a>、<a href="/zh-CN/docs/Web/CSS/computed_value" title="计算值">计算</a>、<a href="/zh-CN/docs/Web/CSS/resolved_value" title="解析值">解析</a>、<a href="/zh-CN/docs/Web/CSS/specified_value" title="指定值">指定</a>、<a href="/zh-CN/docs/Web/CSS/used_value" title="使用值">使用</a>和<a href="/zh-CN/docs/Web/CSS/actual_value" title="实际值">实际</a>值。
+ Definitions of <a href="/zh-CN/docs/Web/CSS/Value_definition_syntax" title="值定义语法">值语法</a>、<a href="/zh-CN/docs/Web/CSS/Shorthand_properties" title="简写属性">简写属性</a>和<a href="/zh-CN/docs/Web/CSS/Replaced_element" title="可替换元素">可替换元素</a>。<%
+        break;
```
Comment (jwhitlock) [inline macros/CSS_key_concepts.ejs]: 👍 tested on [Web/CSS/initial_value](https://developer.mozilla.org/zh-CN/docs/Web/CSS/initial_value). It also works in French, despite GitHub not liking the format...

**Comment by jwhitlock** [comment]
I've merged the macros that I've reviewed so far as PR #156. Next macro is ``ChapterTOC``.  I'm getting more excited to add gettext support.

**Comment by BychekRU** [comment]
Hello everyone!
It's so sad to see that we're so slow with this. I think it's not good when we can't apply localization updates more than 2 months!
Of course I understand when we test macros like `AppsLandingRightSide.ejs` because typos in such macros can't break markup for localized version of site or even much more. But I really can't get what to review in macros like `ArticlesByTag.ejs` l10n in them is done easy no html markup, no JavaScript, localizer can't break anything, only one string `"locale":"translation"`. GitHub highlights JS syntax, fast & easy to check localazer's work.
Also I don't like splitting localization updates to small PRs, I don't think that it must have so much attention. Before we, key contributors, were fast to submit our changes to l10n, I thought it'll become even better to manage macros in GitHub, but now I'm very dissapointed.
Please make something with l10n updates! Maybe to assign someone or please tell me who manages l10n of MDN and macros? Is it better to create bug in Bugzilla or we already have this one?

**Comment by jwhitlock** [comment]
@BychekRU your criticism is valid, it is sad that localizations take this long to merge.  However, we've done quite a bit over the last few months to make it easier:

PR #152 added an EJS linter for macros, and fixed some existing invalid macros.  This runs in TravisCI and detects a class of issues in PRs without manual testing.  This gives us the confidence to visually inspect some changes.  However, this PR was submitted before that linter, and hasn't been checked with it.  A rebase is needed to be able to run the linter.

Testing a macro currently requires running Kuma with the new macro code, and re-rendering a page using the macro.  One difficult part is finding a page that uses a macro.  PR #149 and https://github.com/mozilla/kuma/pull/4179 added a new [macros dashboard](https://developer.mozilla.org/en-US/dashboards/macros), which combines Kuma's site search and KumaScript's list of loaded macros, to allow reviewers to quickly find macro usage examples (or that a macro is unused).

Finding a page doesn't get it into your testing environment.  In the last few weeks, we merged https://github.com/mozilla/kuma/pull/4248, which add a [scrape_document command](https://kuma.readthedocs.io/en/latest/data.html#add-an-mdn-wiki-document), which lets developers and reviewers get one or more production webpages into their local environment.  This will be useful for macro development and manual review.

This is still not a great way to test or develop translations.  Changing or adding a translation shouldn't be a logic change that requires adding or updating tests.  It shouldn't even require changing the macro! @yfdyh000 suggested using standard translation tools and [Pontoon](https://pontoon.mozilla.org/projects/mdn/) in [bug 1340342](https://bugzilla.mozilla.org/show_bug.cgi?id=1340342), and I think it's a great idea.  This will require changing the process for localization in Kuma as well, moving locales to https://github.com/mozilla-l10n/mdn-l10n. I think it will be late July at the earliest before we can start converting KumaScript macros to use Pontoon.

I was willing to review and merge some translations, but asked for them to be in small PRs, ideally one per macro. @yfdyh000 decided to submit one with all the "just string" changes. OK. It takes 10-15 minutes to review each of them using the manual process, and I was trying to do 3-5 a week.

**Comment by jwhitlock** [comment]
Sorry, slipped and hit "close and comment". Editing my comment.

**Comment by yfdyh000** [comment]
Wish bug 1340342 smooth. Not urgent for the translation landing, maybe.

**Comment by BychekRU** [comment]
@jwhitlock Great news. Thank you for this rich answer. Of course focus on transition l10n for using on Pontoon.

**Comment by jwhitlock** [comment]
Thanks @yfdyh000 and @BychekRU.  And I'm sorry for not updating the PR with the plan two months ago - it would have saved a lot of frustration!

**Pull request closed without merging**